backend/ovs/device.go

```diff
@@ -0,0 +1,661 @@
+	oflowProto      string
+	opHandler       OpHandler
+	clusterNetwork  ip.IP4Net
+	nodeNetwork     ip.IP4Net
+	servicesNetwork ip.IP4Net
+	gwAddr          net.IP
+}
+
+func newOVSDevice() (*ovsDevice, error) {
+	return newOVSDeviceWithHandler(&liveOpHandler{})
+}
+
+func newOVSDeviceWithHandler(handler OpHandler) (*ovsDevice, error) {
+	bridge := &ovsDevice{
+		bridgeName: "br0",
+		oflowProto: "OpenFlow13",
+		opHandler:  handler,
+	}
+	err := bridge.genericSetup()
+	return bridge, err
+}
+
+//*****************************************************************************
+
+type OpHandler interface {
+	OvsExec(cmd string, args ...string) ([]byte, error)
+	Modprobe(module string) ([]byte, error)
+	LinkAdd(link netlink.Link) error
+	LinkDel(link netlink.Link) error
+	LinkByIndex(index int) (netlink.Link, error)
+	LinkByName(name string) (netlink.Link, error)
+	LinkSetUp(link netlink.Link) error
+	LinkSetMaster(link netlink.Link, master *netlink.Bridge) error
+	AddrAdd(link netlink.Link, addr *netlink.Addr) error
+	AddrList(link netlink.Link, family int) ([]netlink.Addr, error)
+	RouteList(link netlink.Link, family int) ([]netlink.Route, error)
+	RouteAdd(route *netlink.Route) error
+	RouteDel(route *netlink.Route) error
+	RouteDelWithProtocol(route *netlink.Route, protocol uint8) error
+	Sysctl(item, value string) error
+}
+
+type liveOpHandler struct{}
+
+func (h *liveOpHandler) OvsExec(cmd string, args ...string) ([]byte, error) {
+	return exec.Command(cmd, args...).CombinedOutput()
+}
+
+func (h *liveOpHandler) Modprobe(module string) ([]byte, error) {
+	return exec.Command("modprobe", module).CombinedOutput()
+}
+
+func (h *liveOpHandler) LinkAdd(link netlink.Link) error {
+	return netlink.LinkAdd(link)
+}
+
+func (h *liveOpHandler) LinkDel(link netlink.Link) error {
+	return netlink.LinkDel(link)
+}
+
+func (h *liveOpHandler) LinkByName(name string) (netlink.Link, error) {
+	return netlink.LinkByName(name)
+}
+
+func (h *liveOpHandler) LinkByIndex(index int) (netlink.Link, error) {
+	return netlink.LinkByIndex(index)
+}
+
+func (h *liveOpHandler) LinkSetUp(link netlink.Link) error {
+	return netlink.LinkSetUp(link)
+}
+
+func (h *liveOpHandler) LinkSetMaster(link netlink.Link, master *netlink.Bridge) error {
+	return netlink.LinkSetMaster(link, master)
+}
+
+func (h *liveOpHandler) AddrAdd(link netlink.Link, addr *netlink.Addr) error {
+	return netlink.AddrAdd(link, addr)
+}
+
+func (h *liveOpHandler) AddrList(link netlink.Link, family int) ([]netlink.Addr, error) {
+	return netlink.AddrList(link, family)
+}
+
+func (h *liveOpHandler) RouteList(link netlink.Link, family int) ([]netlink.Route, error) {
+	return netlink.RouteList(link, family)
+}
+
+func (h *liveOpHandler) RouteAdd(route *netlink.Route) error {
+	return netlink.RouteAdd(route)
+}
+
+func (h *liveOpHandler) RouteDel(route *netlink.Route) error {
+	return netlink.RouteDel(route)
+}
+
+func (h *liveOpHandler) RouteDelWithProtocol(route *netlink.Route, protocol uint8) error {
```

Comment:
Can we get this upstreamed to vishvananda/netlink? Ok, short term to be here but should see if we can move it to netlink lib somehow.